### PR TITLE
Edf fields

### LINKED ISSUE #179: Additional file fields on ExperimentDataFile
**Feature**
In the interest of reproducibility, it would be better to have the input script that is supplied to the cycler for each uploaded raw data file. That way, another researcher can use that input script to reproduce the experimental steps. This file does not have to be parsed in any way.

**Proposed solution**
- [x] Add an `input_script` file field on `ExperimentDataFile`
- [x] This file should be validated - must be a text file (not binary file) but probably can have any extension.
- [x] There should be another (optional) field on the form for "input script" when uploading a data file.
- [x] It makes sense to tackle the uploading of binary files at the same time - #83. This should be another optional field with no parsing needed.
  - [x] Create a new field for this
  - [x] Add a formfield for this (optional)

**Potential blockers**
- [x] What do the files look like? Get an example from the research team.

### PR DESCRIPTION
**Pull request by dandavies99**
This adds two new optional fields to the ExperimentDataFile, one for "settings file" and one for "binary file". These are just `FileFields` as opposed to `UploadedFile` objects because they never need to be parsed or have any of the other properties of data files.

New validators have been created to check the files meet certain criteria. These may need to be reviewed in future as different types of file (with different allowed extensions) come to light. 

closes #179

### CODE REVIEW

**Review by CWestICL — COMMENTED** [review]
When uploading binary files or settings files to existing Experiment Data file object, an error message was thrown saying the uploaded file was invalid. However, going back to the experiment's page showed the files had actually uploaded and could be downloaded.

When uploading binary files or settings files for a new Experiment Data file object, the error was thrown again but recognised the files were invalid and did not create the new Data object.

**Comment by dandavies99** [comment]
Yeah, in general the logic for `UpdateDataFileView` needs checking since all these changes. We chose to not allow users to use that method to change the data file that had been uploaded, which is reflected in the `__init__` method of `NewExperimentDataFileForm`. I can't remember why this was, maybe we don't need to do that now.

So the form is valid, which means the `ExperimentDataFile` object gets saved including the new settings file (`self.object.save()`) but the formset is not valid because... it's not there! The form doesn't show it, just a message saying you can't change the raw data file.

So either, in the view we can remove the check for the formset being valid, or we can choose to allow users to change the file they've uploaded.

File: tests/battDB/test_views.py [hunk 1]
```diff
@@ -647,3 +647,169 @@ def test_upload_view_unparsed_data(self):
...
+
+        # Check ExperimentDataFile has been created
+        edf = bdb.ExperimentDataFile.objects.get(name="Device 4")
+        self.assertTrue(edf.file_exists())
+
+        # Check Experiment Detail view contains message about unparsed data
+        get_response = self.client.get(
+            reverse("battDB:Experiment", kwargs={"pk": self.experiment.id})
+        )
+        self.assertContains(get_response, "Download binary")
```
Comment (AdrianDAlessandro) [inline tests/battDB/test_views.py:696]: Do you mean 
```suggestion
        # Check Experiment Detail view contains button to download binary file
        get_response = self.client.get(
            reverse("battDB:Experiment", kwargs={"pk": self.experiment.id})
        )
        self.assertContains(get_response, "Download binary")
```

File: tests/battDB/test_views.py [hunk 2]
```diff
@@ -647,3 +647,169 @@ def test_upload_view_unparsed_data(self):
...
+
+        # Check ExperimentDataFile has been created
+        edf = bdb.ExperimentDataFile.objects.get(name="Device 5")
+        self.assertTrue(edf.file_exists())
+
+        # Check Experiment Detail view contains message about unparsed data
+        get_response = self.client.get(
+            reverse("battDB:Experiment", kwargs={"pk": self.experiment.id})
+        )
+        self.assertContains(get_response, "settings file")
```
Comment (AdrianDAlessandro) [inline tests/battDB/test_views.py:754]: And
```suggestion
        # Check Experiment Detail view contains button to download settings file
        get_response = self.client.get(
            reverse("battDB:Experiment", kwargs={"pk": self.experiment.id})
        )
        self.assertContains(get_response, "settings file")
```

**Review by AdrianDAlessandro — APPROVED** [review]
Everything seems good now!

File: tests/battDB/test_views.py (hunk 1 above)
Comment (dandavies99) [reply tests/battDB/test_views.py:696]: Looks like it

**Comment by dandavies99** [comment]
That's strange. Maybe because in the `UpdateDataFileView` the `self.object.save()` happens before the file validation?

**Pull request merged**